src/store/slices/testing.slice.ts

```diff
@@ -0,0 +1,206 @@
+
+export const createTestRun = createAsyncThunk("createTestRun", async (payload: {}, thunkApi) => {
+  try {
+    const form = (thunkApi.getState() as RootState).testing.form!;
+    const [, , , owner, repo] = form.repoUrl.split('/');
+    const profile = (thunkApi.getState() as RootState).profile.profile!;
+    const githubToken = (thunkApi.getState() as RootState).session.accessToken || undefined;
+    await thunkApi.dispatch(updateProfile({
+      ...profile, dapp: {
+        ...profile.dapp,
+        name: form.name,
+        subject: form.subject,
+        owner, repo, githubToken
+      }
+    }));
+    await thunkApi.dispatch(clearAccessToken({}));
+
+    const response = await fetch<string>(thunkApi, {
+      method: 'POST', url: '/run', data: form.commitHash,
+      headers: {'Content-type': 'text/plain;charset=utf-8'}
+    });
+    return response.data;
+  } catch (e: any) {
+    return thunkApi.rejectWithValue(e.response.data);
+  }
+});
+
+export const fetchRunStatus = createAsyncThunk("fetchRunStatus", async (payload: {}, thunkApi) => {
+  try {
+    const { uuid, timelineConfig, plannedTestingTasks, unitTestSuccess, hasFailedTasks } = (thunkApi.getState() as RootState).testing;
+    const response = await fetch<RunStatus>(thunkApi, { method: 'GET', url: `/run/${uuid}` });
+    const newTimelineConfig = processTimeLineConfig(response, timelineConfig);
+    const newPlannedTestingTasks = getPlannedTestingTasks(response, plannedTestingTasks);
+
+    const status = response.data.status;
+    const state = response.data.hasOwnProperty('state') && response.data.state ? response.data.state : null;
+    
+    let coverageFile = null;
+    let resultData = null;
+    let runEnded = false;
+    let newUnitTestSuccess = unitTestSuccess;
+    let newHasFailedTasks = hasFailedTasks;
+    let fetching = status === 'certifying' || status === 'building' || status === 'preparing' || status === 'queued' || (status === 'finished' && state === 'running');
+    let refetchMin = status === 'certifying' ? 2 : 5;
+
+    if (status === 'finished') {
+      const isArrayResult = Array.isArray(response.data.result);
+      const resultJson = isArrayResult ? response.data.result[0] : response.data.result;
+      if (isArrayResult) coverageFile = response.data.result[1];
+      resultData = resultJson;
+      runEnded = true;
+      newUnitTestSuccess = processFinishedJson(resultJson);
+      newHasFailedTasks = isAnyTaskFailure(resultData);
+    }
+
+    return {
+      shouldFetchRunStatus: state === 'running' || state === 'passed',
+      timelineConfig: newTimelineConfig,
+      plannedTestingTasks: newPlannedTestingTasks,
+      unitTestSuccess: newUnitTestSuccess,
+      hasFailedTasks: newHasFailedTasks,
+      runStatus: status,
+      runState: state,
+      coverageFile,
+      resultData,
+      runEnded,
+      fetching,
+      refetchMin,
+    };
+  } catch (e: any) {
+    console.log(e);
+    return thunkApi.rejectWithValue(e.response.data);
+  }
```

Comment:
Isn't this too much of logic within slice. Is that alright?

Reply:
I agree, right now we have too much logic that can be simplified. But I think it'd be better to do that refactor in another PR.
In this task my only concern was to keep the component working without introducing too many changes to the internal logic.